crates/fuel-gas-price-algorithm/src/v1.rs

```diff
@@ -151,13 +146,128 @@ pub struct AlgorithmUpdaterV1 {
+///
+/// e.g. if the decrease range is 50, the hold range is 50, and the increase range is 50:
+///
+/// 0<-- decrease range -->50<-- hold range -->100<-- increase range -->150
+///
+/// The current activity determines the behavior of the DA gas price.
+///
+/// For healthy behavior, the activity should be in the `increase` range.
 #[derive(serde::Serialize, serde::Deserialize, Debug, Clone, PartialEq)]
+pub struct L2ActivityTracker {
+    /// The _size_ of the range of the activity buffer that will increase the gas price
+    normal_range_size: u16,
+    /// The _size_ of the range of the activity buffer that will hold the gas price
+    hold_range_size: u16,
+    /// The _size_ of the range of the activity buffer that will decrease the gas price
+    decrease_range_size: u16,
+    /// The current activity of the L2 chain
+    activity: u16,
+    /// The threshold of activity below which the activity will be decreased
+    block_activity_threshold: ClampedPercentage,
+}
+
+/// Designates the intended behavior of the DA gas price based on the activity of the L2 chain
+pub enum DAGasPriceSafetyMode {
+    /// Should increase DA gas price freely
+    Normal,
+    /// Should not increase the DA gas price
+    Hold,
+    /// Should decrease the DA gas price always
+    Decrease,
+}
+
+impl L2ActivityTracker {
+    pub fn new_full(
+        normal_range_size: u16,
+        hold_range_size: u16,
+        decrease_range_size: u16,
+        block_activity_threshold: ClampedPercentage,
+    ) -> Self {
+        let activity = decrease_range_size
+            .saturating_add(hold_range_size)
+            .saturating_add(normal_range_size);
+        Self {
+            normal_range_size,
+            hold_range_size,
+            decrease_range_size,
+            activity,
+            block_activity_threshold,
+        }
+    }
+
+    pub fn new(
+        normal_range_size: u16,
+        hold_range_size: u16,
+        decrease_range_size: u16,
+        activity: u16,
+        block_activity_threshold: ClampedPercentage,
+    ) -> Self {
+        let max_activity = decrease_range_size
+            .saturating_add(hold_range_size)
+            .saturating_add(normal_range_size);
+        let activity = activity.min(max_activity);
+        Self {
+            normal_range_size,
+            hold_range_size,
+            decrease_range_size,
+            activity,
+            block_activity_threshold,
+        }
+    }
+
+    pub fn new_always_normal() -> Self {
+        Self::new_full(1, 0, 0, ClampedPercentage::new(0))
+    }
+
+    pub fn safety_mode(&self) -> DAGasPriceSafetyMode {
+        if self.activity
+            > self
+                .decrease_range_size
+                .saturating_add(self.hold_range_size)
+        {
+            DAGasPriceSafetyMode::Normal
+        } else if self.activity > self.decrease_range_size {
+            DAGasPriceSafetyMode::Hold
+        } else {
+            DAGasPriceSafetyMode::Decrease
+        }
+    }
```

Comment:
I think this should be tested a little better to make sure it chooses the expected "mode" based on given parameters.

Reply:
Yes, seems to be well covered. However, maybe this will version is more clear?

```
        let increase_threshold = self
            .decrease_range_size
            .saturating_add(self.hold_range_size);
        let hold_threshold = self.decrease_range_size;

        if self.activity > increase_threshold {
            return DAGasPriceSafetyMode::Normal;
        }
        if self.activity > hold_threshold {
            return DAGasPriceSafetyMode::Hold;
        }
        DAGasPriceSafetyMode::Decrease
```

We could even precalculate the thresholds in the constructor if we don't plan to change them in runtime.

Reply:
I'm pre-computing them now.